core/src/cluster_info_vote_listener.rs

```diff
@@ -494,47 +575,47 @@ impl ClusterInfoVoteListener {
             }
             let epoch_stakes = epoch_stakes.unwrap();
 
-            // The last vote slot, which is the greatest slot in the stack
-            // of votes in a vote transaction, qualifies for optimistic confirmation.
-            // We cannot count any other slots in this vote toward optimistic confirmation because:
-            // 1) There may have been a switch between the earlier vote and the last vote
-            // 2) We do not know the hash of the earlier slot
-            if slot == last_vote_slot {
+            // We always track the last vote slot for optimistic confirmation. If we have replayed
+            // the same version of last vote slot that is being voted on, then we also track the
+            // other votes in the proposed tower.
+            if slot == last_vote_slot || accumulate_intermediate_votes {
                 let vote_accounts = epoch_stakes.stakes().vote_accounts();
                 let stake = vote_accounts.get_delegated_stake(vote_pubkey);
                 let total_stake = epoch_stakes.total_stake();
+                let Some(hash) = get_hash(slot) else {
+                    // In this case the supposed ancestor of this vote is missing. This can happen
+                    // if the ancestor has been pruned, or if this is a malformed vote. In either case
+                    // we do not track this slot for optimistic confirmation.
+                    continue;
+                };
```

Comment:
we could do this verification, would add some additional overhead.
however an attacker can already submit votes with incorrect bank hashes to gossip so i'm not sure we need to protect against this ancestor case, since we can't protect against the main one.

It would all fall under slashable behavior for malformed votes, as the bank hash for `last_vote_slot` would identify a block without the ancestor in `SlotHashes`